[GEOS-10723] params.extractor Resource

dieterstueken opened the pull request:
[![GEOS-10723](https://badgen.net/badge/JIRA/GEOS-10723/0052CC)](https://osgeo-org.atlassian.net/browse/GEOS-10723)<!-- PR-BADGE: PLEASE DO NOT REMOVE THIS COMMENT -->

Classes from `org.geoserver.params.extractor` frequently use `InputStreams` and `OutputStreams`. Doing so, it is not clear if the producer or the consumer has to close the stream. Using `Supplier<*Stream>` does not help either. This also leads to additional catch/try/resource blocks spread all over the code. This approach uses `org.geoserver.platform.resource.Resource` instead.

This way the catch/try/resource blocks can be kept small and consistent.
This also enables to consider `Type.UNDEFINED` concerning [GEOS-10705].

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [ ] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [ ] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

dieterstueken commented:
A general problem with `Resource.in()` is, that it returns an `InputStream` to the client code which works with it in different and questionable ways. So, I thought about the functional approach again: 
Almost all usages (126!) try to convert the given `Resource` into some object using try/catch/resource with additional if/else conditions if it is a resource at all and what to do else.

All that boilerplate code can easily be avoided by using a generator `Function<InputStream, T>` and finally returning an Optional<T>. Using an Optional<T> also eases all those else{default} blocks by simply using `Optional.orElse(T)`.

I try to demonstrate it on this PR.

dieterstueken commented on src/extension/params-extractor/src/main/java/org/geoserver/params/extractor/EchoParametersDao.java:
```diff
@@ -50,27 +49,40 @@ public static String getTmpEchoParametersPath() {
+     * @return a list of EchoParameter or an empty list if the resource does not exist.
+     */
+    public static List<EchoParameter> getEchoParameters(Resource resource) {
+        return resource.optionalRead(EchoParametersDao::read).orElseGet(ArrayList::new);
```
this looks much clearer than the deeply cascaded if/try blocks before.

jodygarnett replied:
This makes a strong case for your api change I like it

dieterstueken commented on src/extension/params-extractor/src/main/java/org/geoserver/params/extractor/EchoParametersDao.java:
```diff
@@ -83,52 +95,44 @@ public static void saveOrUpdateEchoParameter(EchoParameter echoParameter) {
 
     public static void saveOrUpdateEchoParameter(
             EchoParameter echoParameter, Resource input, Resource output) {
-        try (InputStream inputStream = input.in();
```
combining input and output will also leave an empty output file if reading fails. Better to separate this. It's also easier to read.

dieterstueken commented on src/extension/params-extractor/src/test/java/org/geoserver/params/extractor/TestSupport.java:
```diff
@@ -59,18 +58,12 @@ private void deleteTestDirectoryQuietly() {
         }
     }
 
-    protected static void doWork(String resourcePath, Consumer<InputStream> consumer)
```
simplified this a bit. This avoids too excessive lambda expressions and significantly improves the readability.

OK, not strictly related to the original topic; split this off into a separate PR?

jodygarnett replied:
I think it is fine here; it is part of your theme of trying to make resource access less difficult; and TestSupport is not an api change for the web application

aaime commented:
One legit QA failure:

```
2022-10-28T14:25:43.0874191Z 14:25:43,087 [INFO] PMD Failure: org.geoserver.platform.resource.Resource:8 Rule:wildcards Priority:3 No Wildcard Imports.
```

(wildcard imports are disallowed by the Google Java coding convention we adopted).

Besides that, the changes to the params-extractor module look good to me, but I'd like to have @jodygarnett review the changes to Resource. Thinking out loud, it looks like a good addition, but it's also asymmetric, probably the same methods should be offered on the write side? Anyways, let's hear from @jodygarnett

dieterstueken commented:
Yes, sorry.

I have to convince my IDE (Intellj)  not to do so.
And I’m too impatient in pushing 😊
I may try to find a spotless plugin or something.
Currently my IDE takes another 7 minutes to compile everything  again after a spoltless:apply.
And if I’m unheeding it bricks it again. Hope I can improve this soon.

Dieter.
Von: Andrea Aime ***@***.***>
Gesendet: Montag, 31. Oktober 2022 12:28
An: geoserver/geoserver ***@***.***>
Cc: Dieter Stüken - con terra GmbH ***@***.***>; Author ***@***.***>
Betreff: Re: [geoserver/geoserver] [GEOS-10723] params.extractor Resource (PR #6304)


One legit QA failure:

2022-10-28T14:25:43.0874191Z 14:25:43,087 [INFO] PMD Failure: org.geoserver.platform.resource.Resource:8 Rule:wildcards Priority:3 No Wildcard Imports.

(wildcard imports are disallowed by the Google Java coding convention we adopted).

Besides that, the changes to the params-extractor module look good to me, but I'd like to have @jodygarnett<https://eur04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fjodygarnett&data=05%7C01%7C%7Cc1b7c2c43fbc40fd257d08dabb32ebe4%7C6e0bfede3fcb4518a16565dc14fe5620%7C0%7C0%7C638028124624166419%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=mkFbjIlaE65ed%2BbhwVskZrraFYaW8eA6Nhq3gn%2BEUKE%3D&reserved=0> review the changes to Resource. Thinking out loud, it looks like a good addition, but it's also asymmetric, probably the same methods should be offered on the write side? Anyways, let's hear from @jodygarnett<https://eur04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fjodygarnett&data=05%7C01%7C%7Cc1b7c2c43fbc40fd257d08dabb32ebe4%7C6e0bfede3fcb4518a16565dc14fe5620%7C0%7C0%7C638028124624322646%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=j7GdhSXsKsnkAsq7gMr8rEHMvEgomRqjDjrK8leZRTw%3D&reserved=0>

—
Reply to this email directly, view it on GitHub<https://eur04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fgeoserver%2Fgeoserver%2Fpull%2F6304%23issuecomment-1296951523&data=05%7C01%7C%7Cc1b7c2c43fbc40fd257d08dabb32ebe4%7C6e0bfede3fcb4518a16565dc14fe5620%7C0%7C0%7C638028124624322646%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=732gbI7%2BbXRXQY1iWqZ2JvqzmMhpjEPairjInpyaAX4%3D&reserved=0>, or unsubscribe<https://eur04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAB3RARRCG77J3YCEG66UTH3WF6UKXANCNFSM6AAAAAARPXYWOI&data=05%7C01%7C%7Cc1b7c2c43fbc40fd257d08dabb32ebe4%7C6e0bfede3fcb4518a16565dc14fe5620%7C0%7C0%7C638028124624322646%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=nhQPoE7cDBy3hmwNk%2BVQ2SEBF53UL6CaKRtv2DPW33M%3D&reserved=0>.
You are receiving this because you authored the thread.Message ID: ***@***.******@***.***>>

jodygarnett commented on src/extension/params-extractor/src/main/java/org/geoserver/params/extractor/RulesDao.java:
```diff
@@ -43,40 +43,53 @@ public static String getTempRulesPath() {
-            }
+    /**
+     * Read a list of Rule from a resource. Return an empty list if the resource does not exist.
+     * This prevents Resource.in() from creating the file or throwing an exception.
```
This comment is no is not part of the api; if you wish to added it below you can.
```suggestion
```

jodygarnett commented on src/extension/params-extractor/src/main/java/org/geoserver/params/extractor/RulesDao.java:
```diff
@@ -43,40 +43,53 @@ public static String getTempRulesPath() {
+     * @param resource to read.
+     * @return a list of Rule or an empty list if the resource does not exist.
+     */
+    public static List<Rule> getRules(Resource resource) {
```
```suggestion
    public static List<Rule> getRules(Resource resource) {
        // This prevents Resource.in() from creating the file or throwing an exception.
```

jodygarnett commented on src/platform/src/main/java/org/geoserver/platform/resource/Resource.java:
```diff
@@ -197,6 +196,44 @@ default byte[] getContents() throws IOException {
+    /**
+     * Read and convert this resource into a result T using a converter function. The Reader
+     * function may (but does not need to) throw an IOException. Returns null if this resource does
+     * not exist. Reading a {@code TYPE.DIRECTORY} resource will raise an IOException.
```
Thanks for writing clear javadocs; consider adding a small code example here

jodygarnett reviewed (commented):
Some small feedback; I rally like the change and clear style of programming it encourages.

dieterstueken commented:
Pushed the suggested cleanup.

> @aaime: ... but it's also asymmetric, probably the same methods should be offered on the write side?

Yes, I think the same. But this may better be solved in context of [GEOS-10705] in general, as it is not related to `params.extractor`. After this gets pushed, I can continue with some other issues related to `.in()`.

aaime commented:
@dieterstueken we try not to have work done half-way, because the other half is not guaranteed to show up. Could you also add the methods for the out side?

dieterstueken commented:
Ok,
I added a simple `write` method using an `IOConsumer` and rewrote `setContents` to demonstrate its usage similar to `getContents`.

But I am not too happy about by now. It would be better to have some functional interface like:
`void write(T data, OutputStream out) throws IOException;`. But this is missing on `org.apache.commons.io`.

I have additional suggestions about this, but it involves some more extensive refactoring going beyond the topic of this issue.

jodygarnett commented:
> Ok, I added a simple `write` method using an `IOConsumer` and rewrote `setContents` to demonstrate its usage similar to `getContents`.
> 
> But I am not too happy about by now. It would be better to have some functional interface like: `void write(T data, OutputStream out) throws IOException;`. But this is missing on `org.apache.commons.io`.
> 
> I have additional suggestions about this, but it involves some more extensive refactoring going beyond the topic of this issue.

Would it help to create your own interface? I have a small hesitation about using org.apache.commons.io interfaces directly in our API. I know we use such libraries internally; but it is different including them on api boundaries.

aaime commented:
@dieterstueken that's why we strongly suggest to have a single focus on a single PR. In this one you are already working on two.
Options:
- Roll back the changes to FileResource and just focus on the params extractor module change
- Create a separate commit for the changes to FileReosurce and make that complete

The preferred approach is the first, but since you just recently started to contributed, I'm happy to go with the second too. Just not happy to merge a work done half-way.

dieterstueken commented:
> @dieterstueken that's why we strongly suggest to have a single focus on a single PR.

You are right. I was tempted to introduce the functional approach within this PR, which was possibly off topic.
I may try to focus on the main aspect again without changing Resource at first.
Is it OK to perform a rebase, even if the comments on it by @jodygarnett may vanish then?

After the original approach was finished, I can start again introducing the functional approach as a separate PR.
Since I'm just coming home drunken, I have to shift it to tomorrow🍺

> ... since you just recently started to contribute ...

yes, I'm still on the learning curve and to impatient, too.

aaime commented:
@dieterstueken there is a way to keep both:
- Close this PR, without further modification
- Create a new clean one, with focused changes

You can refer to this one from newer PRs and nothing is lost. Just an idea!

dieterstueken commented:
This pull request became too expansive concerning functional extensions of `Resource`.
So, it is closed and replaced by [#6331](https://github.com/geoserver/geoserver/pull/6331).
The aspects of functional extensions of `Resource` will then go to a separate issue.
